lsp-headerline.el

```diff
@@ -182,88 +182,95 @@ PATH is the current folder to be checked."
-                            symbol2-name))))
-                   (lsp-headerline--symbol-with-action symbol-to-append full-symbol-2)))
-               enumerated-symbols-hierarchy))))
+          (mapconcat (-lambda (((symbol-to-append &as &DocumentSymbol :deprecated? :name)
+                                . index))
+                       (let* ((symbol2-name
+                               (propertize name
+                                           'font-lock-face
+                                           (if deprecated?
+                                               'lsp-headerline-breadcrumb-deprecated-face
+                                             'lsp-headerline-breadcrumb-symbols-face)))
+                              (symbol2-icon
+                               (lsp-headerline--symbol-icon symbol-to-append))
+                              (full-symbol-2
+                               (concat
+                                (if lsp-headerline-breadcrumb-enable-symbol-numbers
+                                    (concat
+                                     (propertize (number-to-string index)
+                                                 'face
+                                                 'lsp-headerline-breadcrumb-symbols-face)
+                                     " ")
+                                  "")
+                                (if symbol2-icon
+                                    (concat symbol2-icon symbol2-name)
+                                  symbol2-name))))
+                         (lsp-headerline--symbol-with-action symbol-to-append full-symbol-2)))
+                     enumerated-symbols-hierarchy
+                     (format " %s " (lsp-headerline--arrow-icon)))
+        "")
+    ""))
 
 (defun lsp-headerline--build-string ()
   "Build the header-line string."
-  (concat
-   (format "%s " (lsp-headerline--arrow-icon))
-   (s-join
-    (format " %s "
-            (lsp-headerline--arrow-icon))
-    (-flatten
-     (-map
-      (lambda (segment)
-        (pcase segment
-          ('project (lsp-headerline--build-project-component-list))
-          ('file (lsp-headerline--build-file-component-list))
-          ('path-up-to-project
-           (or lsp-headerline--path-up-to-project-component-list
-               (lsp-headerline--build-path-up-to-project-component-list)))
-          ('symbols (lsp-headerline--build-symbol-component-list))))
-      lsp-headerline-breadcrumb-segments)))))
+  (string-trim-right
+   (mapconcat
+    (lambda (segment)
+      (let ((segment-string
+             (pcase segment
+               ('project (lsp-headerline--build-project-string))
+               ('file (lsp-headerline--build-file-string))
+               ('path-up-to-project
+                (or lsp-headerline--path-up-to-project-string
+                    (lsp-headerline--build-path-up-to-project-string)))
+               ('symbols (lsp-headerline--build-symbol-string))
+               (_ ""))))
```

Comment:
maybe user-error here?

Reply:
How about `message`? The only thing I don't like about `user-error` is that the headerline won't update until the user fixes the error.

Reply:
or `lsp-log`